src/core/meshcop/energy_scan_client.cpp

```diff
@@ -0,0 +1,239 @@
+    Ip6::MessageInfo messageInfo;
+    Message *message;
+
+    header.Init();
+    header.SetVersion(1);
+    header.SetType(aAddress.IsMulticast() ? Coap::Header::kTypeNonConfirmable : Coap::Header::kTypeConfirmable);
+    header.SetCode(Coap::Header::kCodePost);
+    header.SetMessageId(0);
+    header.SetToken(NULL, 0);
+    header.AppendUriPathOptions(OPENTHREAD_URI_ENERGY_SCAN);
+    header.AppendContentFormatOption(Coap::Header::kApplicationOctetStream);
+    header.Finalize();
+
+    VerifyOrExit((message = mSocket.NewMessage(0)) != NULL, error = kThreadError_NoBufs);
+    SuccessOrExit(error = message->Append(header.GetBytes(), header.GetLength()));
+
+    sessionId.Init();
+    sessionId.SetCommissionerSessionId(mNetif.GetCommissioner().GetSessionId());
+    SuccessOrExit(error = message->Append(&sessionId, sizeof(sessionId)));
+
+    channelMask.Init();
+    channelMask.SetLength(sizeof(channelMaskBuf));
+    SuccessOrExit(error = message->Append(&channelMask, sizeof(channelMask)));
+
+    channelMaskEntry.SetChannelPage(0);
+    channelMaskEntry.SetMaskLength(sizeof(aChannelMask));
+
+    for (size_t i = 0; i < sizeof(aChannelMask); i++)
+    {
+        channelMaskBuf[sizeof(MeshCoP::ChannelMaskEntry) + i] = (aChannelMask >> (8 * i)) & 0xff;
+    }
+
+    SuccessOrExit(error = message->Append(channelMaskBuf, sizeof(channelMaskBuf)));
+
+    count.Init();
+    count.SetCount(aCount);
+    SuccessOrExit(error = message->Append(&count, sizeof(count)));
+
+    period.Init();
+    period.SetPeriod(aPeriod);
+    SuccessOrExit(error = message->Append(&period, sizeof(period)));
+
+    scanDuration.Init();
+    scanDuration.SetScanDuration(aScanDuration);
+    SuccessOrExit(error = message->Append(&scanDuration, sizeof(scanDuration)));
+
+    memset(&messageInfo, 0, sizeof(messageInfo));
+    messageInfo.GetPeerAddr() = aAddress;
+    messageInfo.mPeerPort = kCoapUdpPort;
+    messageInfo.mInterfaceId = mNetif.GetInterfaceId();
+    SuccessOrExit(error = mSocket.SendTo(*message, messageInfo));
+
+    otLogInfoMeshCoP("sent energy scan query\r\n");
+
+    mCallback = aCallback;
+    mContext = aContext;
+
+exit:
+
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+
+    return error;
+}
+
+void EnergyScanClient::HandleReport(void *aContext, Coap::Header &aHeader, Message &aMessage,
+                                    const Ip6::MessageInfo &aMessageInfo)
+{
+    static_cast<EnergyScanClient *>(aContext)->HandleReport(aHeader, aMessage, aMessageInfo);
+}
+
+void EnergyScanClient::HandleReport(Coap::Header &aHeader, Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    OT_TOOL_PACKED_BEGIN
+    struct
+    {
+        MeshCoP::ChannelMaskTlv tlv;
+        MeshCoP::ChannelMaskEntry header;
+        uint32_t mask;
+    } OT_TOOL_PACKED_END channelMask;
+
+    OT_TOOL_PACKED_BEGIN
+    struct
+    {
+        MeshCoP::EnergyListTlv tlv;
+        uint8_t list[64];
```

Comment:
Not sure is it better to define a macro in openthread-core-config.h, instead of hardcode here and kMaxResults?